File: src/main/java/com/itszuvalex/femtocraft/industry/tiles/TileEntityBaseEntityMicroFurnace.java
/*******************************************************************************
 * Copyright (C) 2013  Christopher Harris (Itszuvalex)
 * Itszuvalex@gmail.com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 ******************************************************************************/

package com.itszuvalex.femtocraft.industry.tiles;

import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;
import com.itszuvalex.femtocraft.FemtocraftGuiHandler;
import com.itszuvalex.femtocraft.industry.blocks.BlockMicroFurnace;
import com.itszuvalex.femtocraft.managers.research.EnumTechLevel;
import com.itszuvalex.femtocraft.utils.FemtocraftDataUtils.Saveable;
import net.minecraft.inventory.ISidedInventory;
import net.minecraft.item.ItemStack;
import net.minecraft.item.crafting.FurnaceRecipes;

public class TileEntityBaseEntityMicroFurnace extends
        TileEntityBaseEntityIndustry implements ISidedInventory {
    public static int powerToCook = 40;
    public static int ticksToCook = 100;
    public static int maxSmelt = 1;
    /**
     * The number of ticks that the current item has been cooking for
     */
    public
    @Saveable
    int furnaceCookTime = 0;
    public
    @Saveable
    int currentPower = 0;
    public
    @Saveable
    ItemStack smeltingStack = null;
    /**
     * The ItemStacks that hold the items currently being used in the furnace
     */
    protected
    @Saveable
    ItemStack[] furnaceItemStacks = new ItemStack[2];
    private
    @Saveable
    String field_94130_e;

    public TileEntityBaseEntityMicroFurnace() {
        super();
        setMaxStorage(800);
        setTechLevel(EnumTechLevel.MICRO);
    }

    @Override
    public int getGuiID() {
        return FemtocraftGuiHandler.MicroFurnaceGuiID;
    }

    protected int getMaxSimultaneousSmelt() {
        return maxSmelt;
    }

    protected int getTicksToCook() {
        // TODO: Load from configs
        return ticksToCook;
    }

    protected int getPowerToCook() {
        // TODO: Load from configs
        return powerToCook;
    }

    /**
     * Returns the number of slots in the inventory.
     */
    @Override
    public int getSizeInventory() {
        return this.furnaceItemStacks.length;
    }

    /**
     * Returns the stack in slot i
     */
    @Override
    public ItemStack getStackInSlot(int par1) {
        return this.furnaceItemStacks[par1];
    }

    /**
     * Removes from an inventory slot (first arg) up to a specified number (second arg) of items and returns them in a
     * new stack.
     */
    @Override
    public ItemStack decrStackSize(int par1, int par2) {
        if (this.furnaceItemStacks[par1] != null) {
            ItemStack itemstack;

            if (this.furnaceItemStacks[par1].stackSize <= par2) {
                itemstack = this.furnaceItemStacks[par1];
                this.furnaceItemStacks[par1] = null;
                return itemstack;
            } else {
                itemstack = this.furnaceItemStacks[par1].splitStack(par2);

                if (this.furnaceItemStacks[par1].stackSize == 0) {
                    this.furnaceItemStacks[par1] = null;
                }

                return itemstack;
            }
        } else {
            return null;
        }
    }

    /**
     * When some containers are closed they call this on each slot, then drop whatever it returns as an EntityItem -
     * like when you close a workbench GUI.
     */
    @Override
    public ItemStack getStackInSlotOnClosing(int par1) {
        if (this.furnaceItemStacks[par1] != null) {
            ItemStack itemstack = this.furnaceItemStacks[par1];
            this.furnaceItemStacks[par1] = null;
            return itemstack;
        } else {
            return null;
        }
    }

    /**
     * Sets the given item stack to the specified slot in the inventory (can be crafting or armor sections).
     */
    @Override
    public void setInventorySlotContents(int par1, ItemStack par2ItemStack) {
        this.furnaceItemStacks[par1] = par2ItemStack;

        if (par2ItemStack != null
            && par2ItemStack.stackSize > this.getInventoryStackLimit()) {
            par2ItemStack.stackSize = this.getInventoryStackLimit();
        }
    }

    /**
     * Returns the name of the inventory.
     */
    @Override
    public String getInvName() {
        return this.isInvNameLocalized() ? this.field_94130_e : "MicroFurnace";
    }

    /**
     * If this returns false, the inventory name will be used as an unlocalized name, and translated into the player's
     * language. Otherwise it will be used directly.
     */
    @Override
    public boolean isInvNameLocalized() {
        return this.field_94130_e != null && this.field_94130_e.length() > 0;
    }

    public void func_94129_a(String par1Str) {
        this.field_94130_e = par1Str;
    }

    /**
     * Returns the maximum stack size for a inventory slot. Seems to always be 64, possibly will be extended. *Isn't
     * this more of a set than a get?*
     */
    @Override
    public int getInventoryStackLimit() {
        return 64;
    }

    @SideOnly(Side.CLIENT)
    /**
     * Returns an integer between 0 and the passed value representing how close the current item is to being completely
     * cooked
     */
    public int getCookProgressScaled(int par1) {
        if (getTicksToCook() == 0) {
            return 0;
        }
        return this.furnaceCookTime * par1 / getTicksToCook();
    }

    @Override
    public boolean isWorking() {
        return smeltingStack != null;
    }

    @Override
    protected boolean canStartWork() {
        if (this.furnaceItemStacks[0] == null) {
            return false;
        }
        if (smeltingStack != null) {
            return false;
        } else if (getCurrentPower() < getPowerToCook()) {
            return false;
        } else {
            ItemStack itemstack = FurnaceRecipes.smelting().getSmeltingResult(
                    this.furnaceItemStacks[0]);
            if (itemstack == null) {
                return false;
            }
            if (this.furnaceItemStacks[1] == null) {
                return true;
            }
            if (!this.furnaceItemStacks[1].isItemEqual(itemstack)) {
                return false;
            }
            int result = furnaceItemStacks[1].stackSize + itemstack.stackSize;
            return result <= getInventoryStackLimit() && result <= itemstack
                    .getMaxStackSize();
        }
    }

    @Override
    protected void startWork() {
        this.smeltingStack = this.furnaceItemStacks[0].copy();
        this.smeltingStack.stackSize = 0;

        int i = 0;
        do {

            if (furnaceItemStacks[1] != null
                && ((smeltingStack.stackSize + i) > furnaceItemStacks[1]
                    .getMaxStackSize())) {
                break;
            }

            if (furnaceItemStacks[0] == null) {
                break;
            }

            if (!consume(getPowerToCook())) {
                break;
            }

            ++smeltingStack.stackSize;
            --this.furnaceItemStacks[0].stackSize;

            if (this.furnaceItemStacks[0].stackSize <= 0) {
                this.furnaceItemStacks[0] = null;
            }
        } while (++i < getMaxSimultaneousSmelt());

        updateBlockState(true);

        this.onInventoryChanged();
        furnaceCookTime = 0;
    }

    @Override
    protected void continueWork() {
        ++furnaceCookTime;
    }

    @Override
    protected boolean canFinishWork() {
        return furnaceCookTime == getTicksToCook();
    }

    @Override
    protected void finishWork() {
        ItemStack itemstack = FurnaceRecipes.smelting().getSmeltingResult(
                this.smeltingStack);

        if (itemstack != null) {
            if (this.furnaceItemStacks[1] == null) {
                this.furnaceItemStacks[1] = itemstack.copy();
                this.furnaceItemStacks[1].stackSize = smeltingStack.stackSize;
            } else if (this.furnaceItemStacks[1].isItemEqual(itemstack)) {
                furnaceItemStacks[1].stackSize += smeltingStack.stackSize;
            }

            smeltingStack = null;
            updateBlockState(false);
            this.onInventoryChanged();
        }

        furnaceCookTime = 0;
    }

    protected void updateBlockState(boolean working) {
        BlockMicroFurnace.updateFurnaceBlockState(working, this.worldObj,
                this.xCoord, this.yCoord, this.zCoord);
    }

    @Override
    public boolean hasGUI() {
        return true;
    }

    @Override
    public void openChest() {
    }

    @Override
    public void closeChest() {
    }


    /**
     * ******************************************************************************** This function is here for
     * compatibilities sake, Modders should Check for Sided before ContainerWorldly, Vanilla Minecraft does not follow
     * the sided standard that Modding has for a while.
     * <p/>
     * In vanilla:
     * <p/>
     * Top: Ores Sides: Fuel Bottom: Output
     * <p/>
     * Standard Modding: Top: Ores Sides: Output Bottom: Fuel
     * <p/>
     * The Modding one is designed after the GUI, the vanilla one is designed because its intended use is for the
     * hopper, which logically would take things in from the top.
     * <p/>
     * This will possibly be removed in future updates, and make vanilla the definitive standard.
     */

    @Override
    public boolean isItemValidForSlot(int i, ItemStack itemstack) {
        switch (i) {
            case (1):
                return false;
            default:
                return true;
        }
    }

    @Override
    public int[] getAccessibleSlotsFromSide(int var1) {
        switch (var1) {
            case (1):
                return new int[]{0};
            case (0):
            case (2):
            case (3):
            case (4):
            case (5):
                return new int[]{1};
            default:
                return new int[]{};
        }
    }

    @Override
    public boolean canInsertItem(int i, ItemStack itemstack, int j) {
        switch (i) {
            case (1):
                return false;
            default:
                return true;
        }
    }

    @Override
    public boolean canExtractItem(int i, ItemStack itemstack, int j) {
        return true;
    }
}

File: src/main/java/com/itszuvalex/femtocraft/industry/tiles/TileEntityBaseEntityMicroFurnace.scala
/*
 * ******************************************************************************
 *  * Copyright (C) 2013  Christopher Harris (Itszuvalex)
 *  * Itszuvalex@gmail.com
 *  *
 *  * This program is free software; you can redistribute it and/or
 *  * modify it under the terms of the GNU General Public License
 *  * as published by the Free Software Foundation; either version 2
 *  * of the License, or (at your option) any later version.
 *  *
 *  * This program is distributed in the hope that it will be useful,
 *  * but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  * GNU General Public License for more details.
 *  *
 *  * You should have received a copy of the GNU General Public License
 *  * along with this program; if not, write to the Free Software
 *  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *  *****************************************************************************
 */
package com.itszuvalex.femtocraft.industry.tiles

import com.itszuvalex.femtocraft.FemtocraftGuiConstants
import com.itszuvalex.femtocraft.api.EnumTechLevel
import com.itszuvalex.femtocraft.api.core.{Configurable, Saveable}
import com.itszuvalex.femtocraft.api.power.PowerContainer
import com.itszuvalex.femtocraft.core.tiles.TileEntityBase
import com.itszuvalex.femtocraft.core.traits.tile.Inventory
import com.itszuvalex.femtocraft.industry.blocks.BlockMicroFurnace
import com.itszuvalex.femtocraft.industry.tiles.TileEntityBaseEntityMicroFurnace._
import com.itszuvalex.femtocraft.industry.traits.IndustryBehavior
import com.itszuvalex.femtocraft.power.traits.PowerConsumer
import com.itszuvalex.femtocraft.utils.BaseInventory
import cpw.mods.fml.relauncher.{Side, SideOnly}
import net.minecraft.item.ItemStack
import net.minecraft.item.crafting.FurnaceRecipes

object TileEntityBaseEntityMicroFurnace {
  @Configurable(comment = "Power tech level.")                          val TECH_LEVEL    = EnumTechLevel.MICRO
  @Configurable(comment = "Power storage maximum.")                     val POWER_STORAGE = 800
  @Configurable(comment = "Power per item to begin processing.")        val POWER_TO_COOK = 40
  @Configurable(comment = "Ticks required to process.")                 val TICKS_TO_COOK = 100
  @Configurable(comment = "Maximum number of items allowed at a time.") val MAX_SMELT     = 1
}

@Configurable class TileEntityBaseEntityMicroFurnace extends TileEntityBase with IndustryBehavior with Inventory with PowerConsumer {
  override def defaultContainer = new PowerContainer(TECH_LEVEL, POWER_STORAGE)

  override def defaultInventory = new BaseInventory(2)

  /**
   * The number of ticks that the current item has been cooking for
   */
  @Saveable var furnaceCookTime          = 0
  @Saveable var currentPower             = 0
  @Saveable var smeltingStack: ItemStack = null
  /**
   * The ItemStacks that hold the items currently being used in the furnace
   */
  @Saveable private var field_94130_e: String = null

  override def hasGUI = true

  override def getGuiID = FemtocraftGuiConstants.MicroFurnaceGuiID


  /**
   * Returns the name of the inventory.
   */
  override def getInventoryName = if (this.hasCustomInventoryName) this.field_94130_e else "MicroFurnace"

  /**
   * If this returns false, the inventory name will be used as an unlocalized name, and translated into the player's
   * language. Otherwise it will be used directly.
   */
  override def hasCustomInventoryName = this.field_94130_e != null && this.field_94130_e.length > 0

  override def isItemValidForSlot(i: Int, itemstack: ItemStack) = i match {
    case (1) =>
      false
    case _ =>
      true
  }

  def func_94129_a(par1Str: String) {
    this.field_94130_e = par1Str
  }

  @SideOnly(Side.CLIENT) def getCookProgressScaled(par1: Int): Int = {
    if (getTicksToCook == 0) {
      return 0
    }
    this.furnaceCookTime * par1 / getTicksToCook
  }

  protected def getTicksToCook = TICKS_TO_COOK

  override def isWorking = smeltingStack != null

  override protected def canStartWork: Boolean = {
    if (getStackInSlot(0) == null) {
      return false
    }
    if (smeltingStack != null) false
    else if (getCurrentPower < getPowerToCook) {
      false
    }
    else {
      val itemstack = FurnaceRecipes.smelting.getSmeltingResult(getStackInSlot(0))
      if (itemstack == null) {
        return false
      }
      if (getStackInSlot(1) == null) {
        return true
      }
      if (!getStackInSlot(1).isItemEqual(itemstack)) {
        return false
      }
      val result = getStackInSlot(1).stackSize + itemstack.stackSize
      result <= getInventoryStackLimit && result <= itemstack.getMaxStackSize
    }
  }

  protected def getPowerToCook = POWER_TO_COOK

  override protected def startWork() {
    smeltingStack = getStackInSlot(0).copy
    smeltingStack.stackSize = 0
    var continue = true
    var i = 0
    do {
      if (getStackInSlot(1) != null && ((smeltingStack.stackSize + i) > getStackInSlot(1).getMaxStackSize)) {
        continue = false
      }
      if (getStackInSlot(0) == null) {
        continue = false
      }
      if (!consume(getPowerToCook)) {
        continue = false
      }
      if (continue) {
        smeltingStack.stackSize += 1
        getStackInSlot(0).stackSize -= 1
        if (getStackInSlot(0).stackSize <= 0) {
          setInventorySlotContents(0, null)
        }
      }
    } while (continue && {i += 1; i} < getMaxSimultaneousSmelt)
    updateBlockState(true)
    this.onInventoryChanged()
    furnaceCookTime = 0
  }

  protected def getMaxSimultaneousSmelt = MAX_SMELT

  protected def updateBlockState(working: Boolean) {
    BlockMicroFurnace.updateFurnaceBlockState(working, this.worldObj, this.xCoord, this.yCoord, this.zCoord)
  }

  override protected def continueWork() {
    furnaceCookTime += 1
  }

  override protected def canFinishWork = furnaceCookTime == getTicksToCook

  override protected def finishWork() {
    val itemstack = FurnaceRecipes.smelting.getSmeltingResult(smeltingStack)
    if (itemstack != null) {
      if (getStackInSlot(1) == null) {
        setInventorySlotContents(1, itemstack.copy)
        getStackInSlot(1).stackSize = smeltingStack.stackSize
      }
      else if (getStackInSlot(1).isItemEqual(itemstack)) {
        getStackInSlot(1).stackSize += smeltingStack.stackSize
      }
      smeltingStack = null
      updateBlockState(false)
      this.onInventoryChanged()
    }
    furnaceCookTime = 0
  }

  override def getAccessibleSlotsFromSide(var1: Int) = var1 match {
    case 1 => Array[Int](0)
    case _ => Array[Int](1)
  }

  override def canInsertItem(i: Int, itemstack: ItemStack, j: Int) = i match {
    case 1 => false
    case _ => true
  }

  override def canExtractItem(i: Int, itemstack: ItemStack, j: Int) = true
}
